m3/reporter.go

```diff
@@ -439,6 +537,62 @@ func (c cachedMetric) ReportTimer(interval time.Duration) {
 	c.reporter.reportCopyMetric(c.metric, c.size, timerType, val, 0)
 }
 
+func (c cachedMetric) ReportSamples(value int64) {
+	c.reporter.reportCopyMetric(c.metric, c.size, counterType, value, 0)
+}
+
+type noopMetric struct {
+}
+
+func (c noopMetric) ReportCount(value int64) {
+}
+
+func (c noopMetric) ReportGauge(value float64) {
+}
+
+func (c noopMetric) ReportTimer(interval time.Duration) {
+}
+
+func (c noopMetric) ReportSamples(value int64) {
+}
+
+type cachedHistogram struct {
+	r                     *reporter
+	name                  string
+	tags                  map[string]string
+	buckets               tally.Buckets
+	cachedValueBuckets    []cachedHistogramBucket
+	cachedDurationBuckets []cachedHistogramBucket
+}
+
+type cachedHistogramBucket struct {
+	valueUpperBound    float64
+	durationUpperBound time.Duration
+	metric             cachedMetric
+}
+
+func (h cachedHistogram) ValueBucket(
+	bucketLowerBound, bucketUpperBound float64,
+) tally.CachedHistogramBucket {
+	for _, b := range h.cachedValueBuckets {
+		if b.valueUpperBound >= bucketUpperBound {
+			return b.metric
+		}
+	}
+	return noopMetric{}
```

Comment:
This is tally itself breaking its own contracts.  I can panic if you'd like.  Otherwise the interface starts to look very not great.